### PR TITLE
[RSDK-869] Allow for servo to turn more than 180

### PR DESCRIPTION
**Pull request by npmenard**
This PR slightly change the gpio servo model to support servo that rotates more than 180 degrees. I needed to change the types in the servo interface because we use to encode servo position in an uint8. Note that this doesn't affect API since the degree value passed in protobuf messages is a uint32

### CODE REVIEW

File: components/servo/gpio/servo.go [hunk 1]
```diff
@@ -193,7 +190,7 @@ func newGPIOServo(ctx context.Context, deps registry.Dependencies, cfg config.Co
 var _ = servo.LocalServo(&servoGPIO{})
 
 // Given minUs, maxUs, deg and frequency attempt to calculate the corresponding duty cycle pct.
-func mapDegToDutyCylePct(minUs, maxUs uint, deg float64, frequency uint) float64 {
+func mapDegToDutyCylePct(minUs, maxUs uint, minDeg, maxDeg, deg float64, frequency uint) float64 {
```
Comment (randhid) [inline components/servo/gpio/servo.go:193]: Should these also be functions on the servo object and use the stored servo variables? Would it be less/more boiler plate? Otherwise looks good to me.

**Review by randhid — APPROVED** [review]

**Comment by randhid** [comment]
Let's test with a few servos in office when you're back.

File: components/servo/gpio/servo.go [hunk 2]
```diff
@@ -133,7 +130,7 @@ func newGPIOServo(ctx context.Context, deps registry.Dependencies, cfg config.Co
 		return nil, errors.Wrap(err, "couldn't get servo pin pwm frequency")
 	}
 	if attr.Frequency != nil {
-		if frequency > 450 || frequency == 0 {
+		if *attr.Frequency > 450 || *attr.Frequency == 0 {
```
Comment (penguinland) [inline components/servo/gpio/servo.go:133]: :eyes: Very good catch here!

File: components/servo/gpio/servo.go [hunk 3]
```diff
@@ -63,15 +63,12 @@ func (config *servoConfig) Validate(path string) ([]string, error) {
 		if config.MaxDeg != nil && *config.StartPos > *config.MaxDeg {
 			return nil, viamutils.NewConfigValidationError(path, errors.New("starting_position_degs cannot be higher than max_angle_deg"))
 		}
-		if *config.StartPos < minDeg || *config.StartPos > maxDeg {
+		if *config.StartPos < defaultMinDeg || *config.StartPos > defaultMaxDeg {
 			return nil, viamutils.NewConfigValidationError(path, errors.New("starting_position_degs should be between 0 and 180"))
```
[the remark this answers is not in the document]
Comment (npmenard) [reply components/servo/gpio/servo.go]: yep correct I changed the logic to take into account set limits

File: components/servo/gpio/servo.go [hunk 4]
```diff
@@ -299,7 +296,7 @@ func (s *servoGPIO) Move(ctx context.Context, ang uint8, extra map[string]interf
 	if angle > s.max {
 		angle = s.max
 	}
-	pct := mapDegToDutyCylePct(s.minUs, s.maxUs, angle, s.frequency)
+	pct := mapDegToDutyCylePct(s.minUs, s.maxUs, s.min, s.max, angle, s.frequency)
```
[the remark this answers is not in the document]
Comment (npmenard) [reply components/servo/gpio/servo.go:302]: they are used to calculate the degree range

File: components/servo/gpio/servo.go [hunk 5]
```diff
@@ -63,15 +63,12 @@ func (config *servoConfig) Validate(path string) ([]string, error) {
...
+		if *config.StartPos < defaultMinDeg || *config.StartPos > defaultMaxDeg {
 			return nil, viamutils.NewConfigValidationError(path, errors.New("starting_position_degs should be between 0 and 180"))
 		}
 	}
-	if config.MinDeg != nil && *config.MinDeg < minDeg {
-		return nil, viamutils.NewConfigValidationError(path, errors.Errorf("min_angle_deg cannot be lower than %f", minDeg))
-	}
-	if config.MaxDeg != nil && *config.MaxDeg > maxDeg {
-		return nil, viamutils.NewConfigValidationError(path, errors.Errorf("max_angle_deg cannot be higher than %f", maxDeg))
+	if config.MinDeg != nil && *config.MinDeg < 0 {
```
[the remark this answers is not in the document]
Comment (npmenard) [reply components/servo/gpio/servo.go:73]: we don't enforce an upper limit for maxDeg since it's dependent on the hardware.  in theory we should also ignore the test minDeg and let a user set a negative minimum deg

File: components/servo/gpio/servo.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (penguinland) [reply components/servo/gpio/servo.go:302]: :facepalm: I misread. Sorry about that!

File: components/servo/gpio/servo.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (penguinland) [reply components/servo/gpio/servo.go:73]: ...so why not do that, too? (i.e., why not get rid of this check and allow a negative minimum value?)

**Review by penguinland — APPROVED** [review]
LGTM!

**Comment by penguinland** [comment]
Does this count as a breaking change? I'm inclined to say no, but I want a second opinion on that.

**Comment by edaniels** [comment]
It's only semantically breaking in the slightest way and only specific to go; you'd have to assume you are expecting errors to happen above 180 and you may be testing/working against that. I think that's an okay breaking change to make. 

API wise, it's not in proto as @npmenard mentioned and widening an integral type is also not breaking.
